### DISCUSSION TITLE
Problems with my UEFI Network Driver built with EDK2

### DISCUSSION THREAD

**Discussion by darinh999** [tianocore/edk2#6414] [Q&A]
Hello all,

I have developed a Network Driver for a proprietary card, and am currently stuck on a problem.  The driver is initialized, functional and able to receive an ethernet packet successfully.  The UEFI then shuts it down and promptly reinitializes it.  I have no idea why.  I am expecting to receive a DHCP IP address assignment packet, or something, I am not sure what the driver should expect to receive.  But it never happens.  Here is the ordering of the Function Table calls:

```
OpCode 0x1  [PXE_OPCODE_START]
OpCode 0x3  [PXE_OPCODE_GET_INIT_INFO]
OpCode 0x5  [PXE_OPCODE_INITIALIZE]
OpCode 0xA  [PXE_OPCODE_STATION_ADDRESS]
OpCode 0x7  [PXE_OPCODE_SHUTDOWN]
OpCode 0x2  [PXE_OPCODE_STOP]
OpCode 0x1  [PXE_OPCODE_START]
OpCode 0x5  [PXE_OPCODE_INITIALIZE]
OpCode 0x9  [PXE_OPCODE_RECEIVE_FILTERS]
OpCode 0x9  [PXE_OPCODE_RECEIVE_FILTERS]
OpCode 0x9  [PXE_OPCODE_RECEIVE_FILTERS]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0xE  [PXE_OPCODE_GET_STATUS]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0x11 [PXE_OPCODE_RECEIVE]
OpCode 0xE  [PXE_OPCODE_GET_STATUS]
```

Is this ordering normal?  Is there something that I did wrong to make it call SHUTDOWN, STOP, START, then INITIALIZE?
If that part is expected, can someone tell me what the UEFI is trying to accomplish with this sequence of GET_STATUS and RECEIVE calls?  The RECEIVE/GET_STATUS sequence continues indefinitely.

I appreciate any insight you can provide.

Thanks!
Darin

**Comment by Nable80** [comment]
@darinh999 

> Is this ordering normal? Is there something that I did wrong to make it call SHUTDOWN, STOP, START, then INITIALIZE?

I think this comment in ``SimpleNetworkDriverStart()`` may resolve your concerns:
https://github.com/tianocore/edk2/blob/master/NetworkPkg/SnpDxe/Snp.c#L637-L647

> The RECEIVE/GET_STATUS sequence continues indefinitely.

This is just a guess, please check that your driver reports media status correctly. It'll also help if you find some way to print stack traces to find which functions are actually calling your handler.
